streaming/src/main/scala/org/apache/gearpump/streaming/dsl/window/impl/WindowRunner.scala

```diff
@@ -19,133 +19,121 @@ package org.apache.gearpump.streaming.dsl.window.impl
           win.intersects(each)
         }
       })
       var mergedWin = win
-      val mergedInputs = FastList.newListWith(input)
+      val mergedInputs = FastList.newListWith(in -> time)
       intersected.forEach(new Procedure[Window] {
         override def value(each: Window): Unit = {
           mergedWin = mergedWin.span(each)
-          mergedInputs.addAll(windowInputs.remove(each))
+          mergedInputs.addAll(winIns.remove(each))
         }
       })
-      windowInputs.put(mergedWin, mergedInputs)
+      winIns.put(mergedWin, mergedInputs)
     }
-
   }
 
-  override def trigger(time: Instant): Unit = {
-    groupedWindowInputs.forEachKeyValue(new Procedure2[GROUP, TreeSortedMap[Window, FastList[IN]]] {
-      override def value(group: GROUP, windowInputs: TreeSortedMap[Window, FastList[IN]]): Unit = {
-        onTrigger(group, windowInputs)
-      }
-    })
-
+  override def trigger(time: Instant): TraversableOnce[(OUT, Instant)] = {
     @annotation.tailrec
-    def onTrigger(group: GROUP, windowInputs: TreeSortedMap[Window, FastList[IN]]): Unit = {
+    def onTrigger(outputs: ArrayBuffer[(OUT, Instant)]): TraversableOnce[(OUT, Instant)] = {
       if (windowInputs.notEmpty()) {
         val firstWin = windowInputs.firstKey
         if (!time.isBefore(firstWin.endTime)) {
           val inputs = windowInputs.remove(firstWin)
-          if (groupedFnRunners.containsKey(group)) {
-            val runner = FunctionRunner.withEmitFn(groupedFnRunners.get(group),
-              (output: OUT) => {
-                taskContext.output(Message(output, time))
-              })
-            val setup = groupedRunnerSetups.get(group)
-            if (!setup) {
-              runner.setup()
-              groupedRunnerSetups.put(group, true)
-            }
-            inputs.forEach(new Procedure[IN] {
-              override def value(t: IN): Unit = {
-                // .toList forces eager evaluation
-                runner.process(t).toList
+          if (!setup) {
+            fnRunner.setup()
+            setup = true
+          }
+          inputs.forEach(new Procedure[(IN, Instant)] {
+            override def value(v: (IN, Instant)): Unit = {
+              fnRunner.process(v._1).foreach {
+                out: OUT => outputs += (out -> v._2)
               }
-            })
-            // .toList forces eager evaluation
-            runner.finish().toList
-            if (groupBy.window.accumulationMode == Discarding) {
-              runner.teardown()
-              groupedRunnerSetups.put(group, false)
-              // dicarding, setup need to be called for each window
-              onTrigger(group, windowInputs)
-            } else {
-              // accumulating, setup is only called for the first window
-              onTrigger(group, windowInputs)
             }
+          })
+          fnRunner.finish().foreach {
+            out: OUT => outputs += (out -> firstWin.endTime.minusMillis(1))
```

Comment:
I'm wondering, does the message time order in `outputs` matter? If so, is it guaranteed?